File: src/test/ironfist/container/creators/InstanceMethodCreatorTest.java
package ironfist.container.creators;

import ironfist.container.InstanceCreator;
import ironfist.container.ObjectRegistry;
import ironfist.container.Resolver;
import ironfist.container.resolvers.ValueResolver;

import java.lang.reflect.InvocationTargetException;
import java.util.Arrays;

import junit.framework.TestCase;

public class InstanceMethodCreatorTest extends TestCase {

  private boolean arg0Called;

  public Object createInstance0Arg() {
    arg0Called = true;
    return this;
  }

  public void test0ArgCreator() throws IllegalArgumentException,
      InstantiationException, IllegalAccessException, InvocationTargetException {
    InstanceCreator creator = new InstanceMethodCreator(new ValueResolver(
      new ObjectRegistry(), this), "createInstance0Arg", new Resolver[0]);
    Object actual = creator.newInstance();
    assertEquals(this, actual);
    assertTrue(arg0Called);
  }

  private boolean arg1Called;
  private Object[] arg1Args;

  public Object createInstance1Arg(String value) {
    arg1Called = true;
    arg1Args = new Object[]{value};
    return this;
  }

  public void test1ArgCreator() throws IllegalArgumentException,
      InstantiationException, IllegalAccessException, InvocationTargetException {
    ObjectRegistry registry = new ObjectRegistry();
    InstanceCreator creator = new InstanceMethodCreator(new ValueResolver(
      registry, this), "createInstance1Arg", new Resolver[]{new ValueResolver(
      registry, "something")});
    Object actual = creator.newInstance();
    assertEquals(this, actual);
    assertTrue(arg1Called);
    assertTrue(Arrays.equals(new Object[]{"something"}, arg1Args));
  }

  private boolean arg2Called;
  private Object[] arg2Args;

  public Object createInstance2Arg(int value) {
    arg2Called = true;
    arg2Args = new Object[]{value};
    return this;
  }

  public void test1ArgIntCreator() throws IllegalArgumentException,
      InstantiationException, IllegalAccessException, InvocationTargetException {
    ObjectRegistry registry = new ObjectRegistry();
    InstanceCreator creator = new InstanceMethodCreator(new ValueResolver(
      registry, this), "createInstance2Arg", new Resolver[]{new ValueResolver(
      registry, "1")});
    Object actual = creator.newInstance();
    assertEquals(this, actual);
    assertTrue(arg2Called);
    assertTrue(Arrays.equals(new Object[]{1}, arg2Args));
  }

}

File: src/test/java/ukalus/container/creators/InstanceMethodCreatorTest.kt
package ukalus.container.creators

import junit.framework.TestCase
import junit.framework.TestCase.*
import ukalus.container.ObjectRegistry
import ukalus.container.Resolver
import ukalus.container.resolvers.ValueResolver
import java.util.*

class InstanceMethodCreatorTest : TestCase() {

    private var arg0Called: Boolean = false

    fun createInstance0Arg(): Any {
        arg0Called = true
        return this
    }

    fun test0ArgCreator() {
        val creator = InstanceMethodCreator(ValueResolver(ObjectRegistry(), this), "createInstance0Arg", arrayOf<Resolver>())
        val actual = creator.newInstance()
        assertEquals(this, actual)
        assertTrue(arg0Called)
    }

    private var arg1Called: Boolean = false
    private var arg1Args: Array<Any>? = null

    fun createInstance1Arg(value: String): Any {
        arg1Called = true
        arg1Args = arrayOf<Any>(value)
        return this
    }

    fun test1ArgCreator() {
        val registry = ObjectRegistry()
        val creator = InstanceMethodCreator(ValueResolver(registry, this), "createInstance1Arg", arrayOf<Resolver>(ValueResolver(registry, "something")))
        val actual = creator.newInstance()
        assertEquals(this, actual)
        assertTrue(arg1Called)
        assertTrue(Arrays.equals(arrayOf<Any>("something"), arg1Args))
    }

    private var arg2Called: Boolean = false
    private var arg2Args: Array<Any>? = null

    fun createInstance2Arg(value: Int): Any {
        arg2Called = true
        arg2Args = arrayOf<Any>(value)
        return this
    }

    fun test1ArgIntCreator() {
        val registry = ObjectRegistry()
        val creator = InstanceMethodCreator(ValueResolver(registry, this), "createInstance2Arg", arrayOf<Resolver>(ValueResolver(registry, "1")))
        val actual = creator.newInstance()
        assertEquals(this, actual)
        assertTrue(arg2Called)
        assertTrue(Arrays.equals(arrayOf<Any>(1), arg2Args))
    }

}
